core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -695,11 +690,16 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       assert (vlen % 8 == 0);
       assert (longArray != null);
 
-      if (numKeys == MAX_CAPACITY
-        // The map could be reused from last spill (because of no enough memory to grow),
-        // then we don't try to grow again if hit the `growthThreshold`.
-        || !canGrowArray && numKeys > growthThreshold) {
-        return false;
+      if (numKeys >= growthThreshold) {
+        if (longArray.size() / 2 == MAX_CAPACITY) {
```

Comment:
We grow the array when `numKeys >= growthThreshold` and `growthThreshold = capacity * 0.5`.  But we actually allocate `capacity * 2` spaces for the array.

So actually `numKeys < growthThreshold = capacity * 0.5 < array length = capacity * 2` should hold true.

Because `numKeys < growthThreshold` is always true, if `numKeys == MAX_CAPACITY`, the capacity would be `MAX_CAPACITY * 2` at least and the length of array will be more than `MAX_CAPACITY * 4`.

But in `allocate`, there is an assert of `capacity <= MAX_CAPACITY`. Looks like those condition are inconsistent.